### PR TITLE
8269752: C2: assert(false) failed: Bad graph detected in build_loop_late

### PR DESCRIPTION
**Pull request by rwestrel**
Some nodes are sunk out of the outer loop of vMeth(). A CastII node is
added on the exit path of the loop to pin the nodes out of the
loop. The dominating test of that CastII is:

if (i >= 132) {

so the CastII is updated with type int [132..]

The inner loop is eliminated, the outer loop is turned into a counted
loop, pre/main/post loops are created. The CastII control is now a
Region that merges paths from the pre/main/post loops. Its data input
is a Phi that merges the iv Phis of the 3 loops.

The loop body includes an early exit:

if (i2 != 0) {
    vMeth_check_sum += i3;
    return;
}

That test is guaranteed to be taken at the second iteration. As a
consequence, the backedge of the main loop is removed. What's left of
the main loop is still guarded by a an entry test that compares the iv
value out of the pre loop with an Opaque1 node. The CastII is now only
reachable through that test. The iv Phi of the pre loop constant folds
to a value that's not >= 132. As a consequence the CastII becomes
top. top propagates to other data node and as a consequence a Phi is
replaced by its only non top input. That causes the assert failure.

The CastII should not be reachable anymore once the iv Phi of the pre
loop has constant folded and the early exit in the main loop is always
taken. But because the test that guards the main loop uses an Opaque1,
it cannot constant fold. As a fix, I propose removing the Opaque1 once
the backedge of the main loop is removed. This allows the entry test
to constant fold. The Opaque1 has no purpose anymore given the loop
doesn't exist at this point.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269752](https://bugs.openjdk.java.net/browse/JDK-8269752): C2: assert(false) failed: Bad graph detected in build_loop_late

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to 52028ab3b1f0b0e72944fff8077daa4e0cd43e71
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/247/head:pull/247` \
`$ git checkout pull/247`

Update a local copy of the PR: \
`$ git checkout pull/247` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/247/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 247`

View PR using the GUI difftool: \
`$ git pr show -t 247`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/247.diff">https://git.openjdk.java.net/jdk17/pull/247.diff</a>

</details>

### CODE REVIEW

**Review by chhagedorn — APPROVED** [review]
Nice analysis, that looks reasonable to me.

**Comment by rwestrel** [comment]
@chhagedorn  thanks for the review

**Review by chhagedorn — APPROVED** [review]

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -4920,28 +4921,32 @@ Node* PhaseIdealLoop::compute_lca_of_uses(Node* n, Node* early, bool verify) {
...
   }
   Node* cmpzm = bolzm->in(1);
   if (cmpzm == NULL || !cmpzm->is_Cmp()) {
-    return false;
+    return NULL;
   }
   // compares can get conditionally flipped
+
+  int input = is_main_loop() ? 2 : 1;
+  bool res = cmpzm->in(input)->Opcode() == Op_Opaque1;
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:4948]: You need to check `input < cmpzm->in(input)` and  `cmpzm->in(input)` for `NULL`. There could be a "dead" path.

Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp:4950]: Thanks for reviewing this. I updated the patch.

**Review by vnkozlov — APPROVED** [review]
Good.

**Pull request closed without merging**